mypyc/ir/rtypes.py

```diff
@@ -820,17 +820,11 @@ def make_simplified_union(items: list[RType]) -> RType:
         items = flatten_nested_unions(items)
         assert items
 
-        # Remove duplicate items using set + list to preserve item order
-        seen = set()
-        new_items = []
-        for item in items:
-            if item not in seen:
-                new_items.append(item)
-                seen.add(item)
-        if len(new_items) > 1:
-            return RUnion(new_items)
+        unique_items = list(dict.fromkeys(items))
+        if len(unique_items) > 1:
+            return RUnion(unique_items)
         else:
-            return new_items[0]
+            return unique_items[0]
```

Comment:
No idea if it actually makes a difference in terms of performance, but I feel like I mildly preferred your earlier idea of only casting it to a `list` if it's actually necessary, i.e.:

```py
        unique_items = dict.fromkeys(items)
        if len(unique_items) > 1:
            return RUnion(list(unique_items))
        else:
            return next(iter(unique_items))
```

Reply:
It seems to indeed have a small impact:
<img width="516" alt="image" src="https://user-images.githubusercontent.com/1755071/233852642-3045ccc9-e4e1-4bff-81c5-3bf973309528.png">

Reply:
So yeah reverted to that